tests/test_interpolation.py

```diff
@@ -753,3 +758,251 @@ def test_none_value_in_quoted_string(restore_resolvers: Any) -> None:
     OmegaConf.register_new_resolver("test", lambda x: x)
     cfg = OmegaConf.create({"x": "${test:'${missing}'}", "missing": None})
     assert cfg.x == "None"
+
+
+@pytest.mark.parametrize(
+    ("cfg", "key", "expected_value", "expected_node_type"),
+    [
+        pytest.param(
+            User(name="Bond", age=SI("${cast:int,'7'}")),
+            "age",
+            7,
+            IntegerNode,
+            id="expected_type",
+        ),
+        pytest.param(
+            # This example specifically test the case where intermediate resolver results
+            # are not of the same type as the key.
+            User(name="Bond", age=SI("${cast:int,${drop_last:${drop_last:7xx}}}")),
+            "age",
+            7,
+            IntegerNode,
+            id="intermediate_type_mismatch_ok",
+        ),
+        pytest.param(
+            User(name="Bond", age=SI("${cast:str,'7'}")),
+            "age",
+            7,
+            IntegerNode,
+            id="convert_str_to_int",
+        ),
```

Comment:
This one is similar to the automatic conversion of ValueNodes I think we want to deprecate.

Reply:
I guess we should have that discussion. The corresponding issue is #459, but right now it's not scheduled for 2.1.

One thing that happened since then is the decision for the `env` resolver to only return strings, where we said that one way to convert such strings to other types would be thanks to the automatic conversion of typed nodes from strings. As a result, I wasn't sure that we really wanted to go forward with #459.

Also, on a more personal note, I'd rather not address #459 myself -- at least not for 2.1 (got enough on my plate at this time).

Reply:
I am not planning to do #459 for 2.1.
I think we just need to add a comment here saying that this relies on automatic conversion from the node.

Reply:
> I think we just need to add a comment here saying that this relies on automatic conversion from the node.

Ah, got it! Done in 4af8eb5